apis/field_error.go

```diff
@@ -88,12 +116,79 @@ func (fe *FieldError) ViaKey(key string) *FieldError {
 	if fe == nil {
 		return nil
 	}
-	return fe.ViaField(fmt.Sprintf("[%s]", key))
+	var newErrs []FieldError
+	for _, e := range fe.getNormalizedErrors() {
+		e.ViaField(fmt.Sprintf("[%s]", key))
+		newErrs = append(newErrs, e)
+	}
+	fe.clear()
+	fe.errors = newErrs
+	return fe
 }
 
 // ViaFieldKey is the short way to chain: err.ViaKey(bar).ViaField(foo)
 func (fe *FieldError) ViaFieldKey(field string, key string) *FieldError {
-	return fe.ViaKey(key).ViaField(field)
+	if fe == nil {
+		return nil
+	}
+	var newErrs []FieldError
+	for _, e := range fe.getNormalizedErrors() {
+		e.ViaKey(key).ViaField(field)
+		newErrs = append(newErrs, e)
+	}
+	fe.clear()
+	fe.errors = newErrs
+	return fe
+}
+
+func (fe *FieldError) getNormalizedErrors() []FieldError {
+	if fe == nil {
+		return []FieldError(nil)
+	}
+	var errors []FieldError
+
+	// if this FieldError is a leaf,
+	if fe.Message != "" {
+		err := FieldError{
+			Message: fe.Message,
+			Paths:   fe.Paths,
+			Details: fe.Details,
+		}
+		errors = append(errors, err)
+	}
+
+	// and then collect all other errors recursively.
+	for _, e := range fe.errors {
+		errors = append(errors, e.getNormalizedErrors()...)
+	}
+
+	return errors
+}
+
+// Also collects errors, returns a new collection of existing errors and new errors.
+func (fe *FieldError) Also(errs ...*FieldError) *FieldError {
+	var newErrs []FieldError
+	if fe == nil {
+		fe = &FieldError{}
+	}
+	newErrs = append(newErrs, fe.getNormalizedErrors()...)
+
+	for _, e := range errs {
+		newErrs = append(newErrs, e.getNormalizedErrors()...)
+	}
+
+	fe.clear()
```

Comment:
I think I'm uncomfortable with the self-modifying FieldError pattern if we have paths where the receiver should become non-nil (like this).  I think we should switch to returning a modified FieldError and leaving the receiver intact.

Reply:
I had a version of this that used .Also to collect the errors internally and it was really neat. I will update to that tomorrow